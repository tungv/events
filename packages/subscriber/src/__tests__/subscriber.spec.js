const kefir = require('kefir');

const { stringify } = JSON;
const getChunks = require('../getChunks');

describe('parser', () => {
  it('should work when data are send in batch ending with double newlines', async () => {
    const chunks = [
      ':ok\n\n',
      `id: 1\nevent: name\ndata: [${stringify({ a: 1 })}]\n\n`,
      `id: 2\nevent: name\ndata: [${stringify({ a: 2 })}]\n\n`,
    ];

    const raw$ = kefir.sequentially(0, chunks);

    const events = await raw$
      .thru(getChunks)
      .scan((a, b) => a.concat(b), [])
      .toPromise();

    expect(events).toMatchSnapshot();
  });

  it('should work when multiple data are send in batch ending with double newlines', async () => {
    const chunks = [
      ':ok\n\n',
      `id: 1\nevent: name\ndata: [{"a": 1}]\n\nid: 2\nevent: name\ndata: []\n\n`,
      `id: 3\nevent: name\ndata: [{"a": 2}]\n\n`,
    ];

    const raw$ = kefir.sequentially(0, chunks);

    const events = await raw$
      .thru(getChunks)
      .scan((a, b) => a.concat(b), [])
      .toPromise();

    expect(events).toMatchSnapshot();
  });

  it('should work when data are truncated', async () => {
    const chunks = [
      ':ok\n\n',
      `id: 1\nevent: name\nda`,
      `ta: [${stringify({ a: 1 })}]\n\n`,
      `id: 2\nevent: name\ndata: [${stringify({ a: 2 })}]\n\n`,
    ];

    const raw$ = kefir.sequentially(0, chunks);

    const events = await raw$
      .thru(getChunks)
      .scan((a, b) => a.concat(b), [])
      .toPromise();

    expect(events).toMatchSnapshot();
  });

  it('should work when data are truncated at the first new line', async () => {
    const chunks = [
      ':ok\n\n',
      `id: 1\nevent: name\ndata: [${stringify({ a: 1 })}]\n`,
      `\nid: 2\nevent: name\ndata: [${stringify({ a: 2 })}]\n\n`,
    ];

    const raw$ = kefir.sequentially(0, chunks);

    const events = await raw$
      .thru(getChunks)
      .scan((a, b) => a.concat(b), [])
      .toPromise();

    expect(events).toMatchSnapshot();
  });

  it('should ignore remaining chunks when stream ends', async () => {
    const chunks = [
      ':ok\n\n',
      `id: 1\nevent: name\ndata: [{"a": 1}]\n\nid: 2\nevent: name\ndata: []\n\n`,
      `id: 3\nevent: name\ndata: [{"a": 2}]\n\nabc: 1`,
    ];

    const raw$ = kefir.sequentially(0, chunks);

    const events = await raw$
      .thru(getChunks)
      .scan((a, b) => a.concat(b), [])
      .toPromise();

    expect(events).toMatchSnapshot();
  });
});
