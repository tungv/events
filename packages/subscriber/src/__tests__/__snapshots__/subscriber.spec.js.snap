// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser should ignore remaining chunks when stream ends 1`] = `
Array [
  Object {
    "data": "[{\\"a\\": 1}]",
    "event": "name",
    "id": "1",
  },
  Object {
    "data": "[]",
    "event": "name",
    "id": "2",
  },
  Object {
    "data": "[{\\"a\\": 2}]",
    "event": "name",
    "id": "3",
  },
]
`;

exports[`parser should work when data are send in batch ending with double newlines 1`] = `
Array [
  Object {
    "data": "[{\\"a\\":1}]",
    "event": "name",
    "id": "1",
  },
  Object {
    "data": "[{\\"a\\":2}]",
    "event": "name",
    "id": "2",
  },
]
`;

exports[`parser should work when data are truncated 1`] = `
Array [
  Object {
    "data": "[{\\"a\\":1}]",
    "event": "name",
    "id": "1",
  },
  Object {
    "data": "[{\\"a\\":2}]",
    "event": "name",
    "id": "2",
  },
]
`;

exports[`parser should work when data are truncated at the first new line 1`] = `
Array [
  Object {
    "data": "[{\\"a\\":1}]",
    "event": "name",
    "id": "1",
  },
  Object {
    "data": "[{\\"a\\":2}]",
    "event": "name",
    "id": "2",
  },
]
`;

exports[`parser should work when multiple data are send in batch ending with double newlines 1`] = `
Array [
  Object {
    "data": "[{\\"a\\": 1}]",
    "event": "name",
    "id": "1",
  },
  Object {
    "data": "[]",
    "event": "name",
    "id": "2",
  },
  Object {
    "data": "[{\\"a\\": 2}]",
    "event": "name",
    "id": "3",
  },
]
`;
